### PR TITLE
[tt-train] Add dummy NoOp optimizer

### PR DESCRIPTION
**Pull request by zbaczewskiTT**
### Problem description
Introduces NoOp (No Operation) optimizer. The purpose of this optimizer is to serve as a dummy for testing and debugging training pipelines. 

### What's changed
- **Added NoOp optimizer** in`tt-train/sources/ttml/optimizers/`
- **Updated the build system** to include `no_op.hpp` and `no_op.cpp`
- **Added tests** in `tt-train/tests/optimizers/no_op_test.cpp`
- **Added option to use NoOp optimizer in nanoGPT** in `tt-train/sources/examples/nano_gpt/main.cpp` and added sample configuration file in `tt-train/configs/training_shakespeare_nanollama3_no_op.yaml`

### Results
Loss comparison of NoOp optimizer with AdamW (batch size = 32)
image

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-quick-trigger.yaml) CI passes (if applicable)
- [ ] [TG demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tt-train/sources/ttml/optimizers/no_op.cpp [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+#include "autograd/autocast_tensor.hpp"
+#include "core/tt_tensor_utils.hpp"
+#include "serialization/serializable.hpp"
+
+namespace ttml::optimizers {
+
+NoOp::NoOp(ttml::serialization::NamedParameters parameters, const NoOpConfig& config) :
+    OptimizerBase(std::move(parameters)), m_config(config) {
+    for (const auto& [name, tensor_ptr] : m_parameters) {
+        if (tensor_ptr->get_requires_grad()) {
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/optimizers/no_op.cpp]: main goal of the PR is to avoid memory overhead
please remove m_theta from it

File: tt-train/sources/ttml/optimizers/no_op.hpp [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+#pragma once
+
+#include "optimizers/optimizer_base.hpp"
+#include "serialization/serializable.hpp"
+
+namespace ttml::optimizers {
+
+struct NoOpConfig {
+    float lr{1e-3F};
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/optimizers/no_op.hpp]: why it needs lr?

File: tt-train/sources/ttml/optimizers/no_op.cpp [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+#include <fmt/format.h>
+
+#include "autograd/autocast_tensor.hpp"
+#include "core/tt_tensor_utils.hpp"
+#include "serialization/serializable.hpp"
+
+namespace ttml::optimizers {
+
+NoOp::NoOp(ttml::serialization::NamedParameters parameters, const NoOpConfig& config) :
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/optimizers/no_op.cpp]: No need for NoOpConfig

**Review by rfurko-tt — CHANGES_REQUESTED** [review]

**Comment by rfurko-tt** [comment]
* How did you test that it works as expected? Include that in PR description 
* Add tests 
* Add option to use NoOpOptimizer in `nano_gpt/main.cpp`

File: tt-train/tests/optimizers/no_op_test.cpp [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        losses.emplace_back(loss_value);
+        loss->backward();
+        optimizer.step();
+        ttml::autograd::ctx().reset_graph();
+    }
+
+    auto current_weight_values = ttml::core::to_vector(model.get_weight()->get_value());
+
+    EXPECT_EQ(initial_weight_values, current_weight_values);
+    EXPECT_GE(losses.back(), losses.front() * 0.95f);
```
Comment (rfurko-tt) [inline tt-train/tests/optimizers/no_op_test.cpp]: why it's not equal?

File: tt-train/sources/examples/nano_gpt/main.cpp [hunk 5]
```diff
@@ -851,19 +854,20 @@ int main(int argc, char **argv) {
...
-    auto select_optimizer =
-        [&model, &adamw_params, &config](bool use_moreh_adamw) -> std::unique_ptr<ttml::optimizers::OptimizerBase> {
+    auto select_optimizer = [&model, &adamw_params, &config]() -> std::unique_ptr<ttml::optimizers::OptimizerBase> {
         if (config.enable_mpi) {
             return std::make_unique<RemoteOptimizer>(
                 get_model_parameters(model), config.num_mh_workers, config.socket_type);
-        } else if (use_moreh_adamw) {
+        } else if (config.use_moreh_adamw) {
             return std::make_unique<ttml::optimizers::MorehAdamW>(get_model_parameters(model), adamw_params);
+        } else if (config.use_no_op) {
```
Comment (rfurko-tt) [inline tt-train/sources/examples/nano_gpt/main.cpp:872]: lets move no op if before moreh_adamw and add distinctive print message to make sure that no one can miss that we are running NoOp optimizer

File: tt-train/sources/ttml/optimizers/no_op.cpp [hunk 6]
```diff
@@ -0,0 +1,47 @@
+// SPDX-FileCopyrightText: (c) 2025 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "no_op.hpp"
+
+#include <fmt/format.h>
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/optimizers/no_op.cpp]: do you need this include?

**Review by rfurko-tt — COMMENTED** [review]
Looks much better, please follow up on new messages and update PR description

File: tt-train/sources/ttml/optimizers/no_op.hpp (hunk 2 above)
Comment (zbaczewskiTT) [reply tt-train/sources/ttml/optimizers/no_op.hpp]: Removed

File: tt-train/tests/optimizers/no_op_test.cpp (hunk 4 above)
Comment (zbaczewskiTT) [reply tt-train/tests/optimizers/no_op_test.cpp]: I had multiple batches in mind, but in this case, you are right it should be equal.
Changed

File: tt-train/sources/examples/nano_gpt/main.cpp (hunk 5 above)
Comment (zbaczewskiTT) [reply tt-train/sources/examples/nano_gpt/main.cpp:872]: Moved and added warning message

File: tt-train/sources/ttml/optimizers/no_op.cpp (hunk 6 above)
Comment (zbaczewskiTT) [reply tt-train/sources/ttml/optimizers/no_op.cpp]: Removed

File: tt-train/tests/optimizers/no_op_test.cpp (hunk 4 above)
Comment (rfurko-tt) [reply tt-train/tests/optimizers/no_op_test.cpp]: thanks

File: tt-train/sources/ttml/optimizers/no_op.hpp [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+    void step() override;
+
+    [[nodiscard]] serialization::StateDict get_state_dict() const override;
+    void set_state_dict(const serialization::StateDict& dict) override;
+
+    [[nodiscard]] size_t get_steps() const override;
+    void set_steps(size_t steps) override;
+
+    [[nodiscard]] float get_lr() const override {
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/optimizers/no_op.hpp]: move to cpp

File: tt-train/sources/ttml/optimizers/no_op.hpp [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+        return 0.0f;
+    }
+
+    void set_lr(float lr) override {
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/optimizers/no_op.hpp]: move to cpp

File: tt-train/tests/optimizers/no_op_test.cpp [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        ttml::autograd::ctx().reset_graph();
+        ttml::autograd::ctx().close_device();
+    }
+};
+
+TEST_F(NoOpFullTest, NoOpTest) {
+    using namespace ttml::ops;
+    ttml::autograd::ctx().set_seed(123);
+    auto* device = &ttml::autograd::ctx().get_device();
+    const size_t batch_size = 32;
```
Comment (rfurko-tt) [inline tt-train/tests/optimizers/no_op_test.cpp]: use 32U if unsigned type is used

**Review by rfurko-tt — APPROVED** [review]
LGTM, a few nits and ready to go

File: tt-train/sources/ttml/optimizers/no_op.hpp (hunk 7 above)
Comment (zbaczewskiTT) [reply tt-train/sources/ttml/optimizers/no_op.hpp]: Moved

File: tt-train/sources/ttml/optimizers/no_op.hpp (hunk 8 above)
Comment (zbaczewskiTT) [reply tt-train/sources/ttml/optimizers/no_op.hpp]: Moved

File: tt-train/tests/optimizers/no_op_test.cpp (hunk 9 above)
Comment (zbaczewskiTT) [reply tt-train/tests/optimizers/no_op_test.cpp]: Changed
